app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,170 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.home.blocklist
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.lib.state.Store
+import org.mozilla.fenix.home.HomeFragmentAction
+import org.mozilla.fenix.home.HomeFragmentState
+import org.mozilla.fenix.home.recentbookmarks.RecentBookmark
+import org.mozilla.fenix.home.recenttabs.RecentTab
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem
+
+class BlocklistMiddleware(
+    coroutineScope: CoroutineScope,
+    private val provider: HomescreenBlocklistProvider
+) : Middleware<HomeFragmentState, HomeFragmentAction> {
+
+    private var blocklist: Set<String> = setOf()
+    private var blocklistLoaded = false
+    private var store: Store<HomeFragmentState, HomeFragmentAction>? = null
+
+    init {
+        provider.blocklist.map { updatedList ->
+            blocklist = updatedList
+            if (!blocklistLoaded) {
+                filterLatestState()
+            }
+            blocklistLoaded = true
+        }.launchIn(coroutineScope)
+    }
+
+    /**
+     * Will filter "Change" actions using the blocklist and use "Remove" actions to update
+     * the blocklist.
+     */
+    override fun invoke(
+        context: MiddlewareContext<HomeFragmentState, HomeFragmentAction>,
+        next: (HomeFragmentAction) -> Unit,
+        action: HomeFragmentAction
+    ) {
+        if (!blocklistLoaded) {
+            store = context.store
+            return
+        }
+        next(handleUpdate(context.state, action))
```

Comment:
This pattern looks fragile to me:
* You are dropping actions on the floor (by not calling `next()`). This is problematic since someone wanted those to be dispatched. :)
* Later code uses `store?`. So that code kind of expects you ran into this block to get a store reference and dropped at least one action.
* There seems to be a race since there's no guarantee the loading and the setting of the reference happens in a specific order? They also seem to be read/set from different threads... which may require `@volatile` or `synchronized` to guarantee visibility.
* Why is `filterLatestState()` needed in the first place? Is it an alternative to filtering upfront?

Reply:
- Good call on the dropped actions. I'll fix that
- The `store?` call from `filterLatestState` should only be called once, when the blocklist loads.
- Is this third bullet in reference to `store`? I hadn't considered that the middleware is `invoked()` from a background thread. Thanks for calling that out. 
- I think the following case is still technically possible:
  1. Initial state is filtered by fragment
  2. State is updated by dispatched action before blocklist is loaded within the middleware
  3. If `filterLatestState` were not called when blocklist was loaded, state would remain unfiltered until the next action was intercepted

I think two options for improving this implementation would be:
1. Block the UI thread until the blocklist has been loaded in the middleware, then use the middleware to filter the initial state directly. This has a higher risk of ANRs and slow loads
2. Remove the blocking call from the fragment and completely rely on `filterLatestState` once the blocklist has been loaded. Add more thread safety. The middleware would require the `Store` or initial state on construction, in case it wasn't captured during a dispatch. Higher risk of blocked items being visible and UI jank if those items are displayed and then removed
3. Switch from `DataStore` to `Settings`, hiding the async IO read but hugely simplifying the workflow.